src/backend/access/transam/xlog.c

```diff
@@ -771,6 +795,19 @@ static WALInsertLockPadded *WALInsertLocks = NULL;
  */
 static ControlFileData *ControlFile = NULL;
 
+#define LAST_WRITTEN_LSN_CACHE_BUCKET 1024 /* blocks = 8Mb */
+
+
+/*
+ * Cache of last written LSN for each relation chunk (hash bucket).
+ * Also to provide request LSN for smgrnblocks, smgrexists there is pseudokey=InvalidBlockId which stores LSN of last
+ * relation metadata update.
+ * Size of the cache is limited by GUC variable lastWrittenLsnCacheSize ("lsn_cache_size"),
+ * pages are replaced using LRU algorithm, based on L2-list.
+ * Access to this cache is protected by 'LastWrittenLsnLock'.
+ */
+static HTAB *lastWrittenLsnCache;
+
```

Comment:
Let's move these next to LastWrittenLsnCacheEntry. It's easier to see how it works when they're all together.

Reply:
Sorry, I do not understand what you mean. `LastWrittenLsnCacheEntry` contains `prev` and 'next' which you have suggested to replaces with  `dlist_node lru_node`. Which `next` do you mean here?

Reply:
Oh I meant, "let's move these *close* to LastWrittenLsnCacheEntry". I didn't mean the field `next`.